### PR TITLE
Make sure ReportOpts only stores valid values.

### PR DESCRIPTION
**Pull request by Xitian9**
This addresses #1338. It's a big api change, so let me know your thoughts.

I'd particularly appreciate checking out how hledger-ui and hledger-web fare. I've checked myself and they seem to work fine, but since I don't normally use these tools I might be missing something.

### CODE REVIEW

**Review by simonmichael — COMMENTED** [review]
Nice work! A few comments:

Would you mind documenting the new overline width argument somewhere, eg in StringFormat's haddock (and for my own understanding, could you show an example of the user-visible effect ?)

ReportOpts now includes the notion of time (current day), it's more of a "report spec" than a pure description of report options. If you want to reuse the report options for a different day, you have to regenerate and overwrite its query_. The change seems to simpify code in most cases so I'm ok with this.

"Valuation" isn't obviously a function (to convert amounts to other amounts on a given day) to me, but I'm ok with that name for now.

Rather than adding the query string to UIState, shall we keep it in ReportOpts alongside query_ ? Analogous to Regexp holding both raw and parsed forms.

**Comment by simonmichael** [comment]
(PS, and of course overall this PR reduces partiality / possibility of error calls after startup, which is great.)

**Comment by Xitian9** [comment]
> Would you mind documenting the new overline width argument somewhere, eg in StringFormat's haddock (and for my own understanding, could you show an example of the user-visible effect ?)

Sure thing,
 
> "Valuation" isn't obviously a function (to convert amounts to other amounts on a given day) to me, but I'm ok with that name for now.

I'm happy to remove the type and just leave the full `Day -> MixedAmount -> MixedAmount` in there.

> Rather than adding the query string to UIState, shall we keep it in ReportOpts alongside query_ ? Analogous to Regexp holding both raw and parsed forms.

I'm a bit hesitant to introduce redundancy which could get out of sync, but perhaps this could be ameliorateb by including a `regenerateReportOpts :: Day -> String -> ReportOpts -> Either String ReportOpts` function, which sets a new date, query string, and parsed `Query`.

**Comment by Xitian9** [comment]
Actually, I'm not sure about moving the query string into ReportOpts.

The lack of proper boolean algebra support for query parsing means there are plenty of Queries that cannot be obtained from a query string. What should be stored in the query string in these cases? This comes up for me fairly frequently, as I define a number of Queries in code, rather than parsing them from strings.

**Comment by simonmichael** [comment]
Good point. Maybe String is one option. 

I felt the redundancy exists anyway in hledger-ui, and probably other places sooner or later, so might as well manage it in one place. But I'm not sure about this either. The naming and purpose of regenerateReportOpts feels non-obvious which is a sign that maybe this data model isn't optimal yet. Maybe it's not a big deal, maybe we'll find clearer names. Let's think a bit more.

File: hledger-lib/Hledger/Reports/MultiBalanceReport.hs [hunk 1]
```diff
@@ -241,7 +238,7 @@ makeReportQuery ropts reportspan
     dateqcons        = if date2_ ropts then Date2 else Date
 
 -- | Make a valuation function for valuating MixedAmounts and a given Day
-makeValuation :: ReportOpts -> Journal -> PriceOracle -> Valuation
+makeValuation :: ReportOpts -> Journal -> PriceOracle -> Day -> MixedAmount -> MixedAmount
```
Comment (simonmichael) [inline hledger-lib/Hledger/Reports/MultiBalanceReport.hs:241]: Maybe parenthesise (Day -> MixedAmount -> MixedAmount) here too, like the others below.

**Comment by simonmichael** [comment]
Brainstorming alternatives: we could keep ReportOpts as a pure description of options, like the other Opts types, and add something new that includes the current date and derived Query, like:

```haskell
data ReportSpec = ReportSpec {
   rsopts :: ReportOpts
  ,rstoday :: Day
  ,rsquery :: Query
}
```

Or perhaps it's just not worth the code noise and attention cost.

**Comment by simonmichael** [comment]
The easier alternative: allow ReportOpts' query string to be Nothing (signifying a Query not currently parseable from command line arguments), and maybe rename regenerateReportOpts ?
```haskell
-- | Set a new today-date, query string, and Query derived from these in this ReportOpts.
-- Or return a query parse error.
reportOptsSetQuery :: Day -> T.Text -> ReportOpts -> Either String ReportOpts
```

**Comment by Xitian9** [comment]
> Brainstorming alternatives: we could keep ReportOpts as a pure description of options, like the other Opts types, and add something new that includes the current date and derived Query, like:

This is more or less the current approach used in `multiBalanceReportWith`, except also with a `Day` extracted. Would there remain a query string in `ReportOpts`?

The hesitation I have with any option that keeps a query string in the same place as the derived `Query` is the need to keep them synced. When the query string is floated out into `UIOpts`, it's at least clear that this is something only for handling the UI, and not for report processing. As far as I know, the only use in keeping the query string around is so it can be modified in hledger-ui. Correct me if I'm wrong. Or do you have in mind a use case where there is a pure description of options, and you need to apply the same description on different days?

**Comment by simonmichael** [comment]
Just trying to guess which formulation will be least hassle and error-prone to work with, yes including when you want to reuse some report options with different "today" dates. Maybe that's rare.

I'm realising that ReportOpts isn't a "pure description of options" even now, the today_ field is used for multiple things: parsing the query string, interpreting partial dates in -b/-e/-p, selecting prices for market valuation, etc.

**Comment by Xitian9** [comment]
I'm actually not 100% clear on what constitutes a pure description of options. Here are some questions I have.

1. Does `today_` belong in `ReportOpts`? If so, what is it and should it be used for?
    a. Valuation date?
    b. Resolving `SmartDate`s parsed from a query string?
    c. Resolving `SmartDate`s parsed from -b/-e/-p
2. There are dates involved in `today_`, `period_`, and `query_`. Which takes priority when all are present? Can some of these be discarded when we have the others?
3. There's some amount of redundancy between `period_` and `interval_`. Should we clarify the relationship?
4. A bunch of flags in `ReportOpts` are later put into a `Query` in some way, including `period_`, `statuses_`, `depth_`, `date2_`, `empty_`, `real_`, and a query string. Should we separate the query from these so it's clear which to use?
5. There are some options which are related to display on a cli, including `pretty_tables_`, `color_`, and `transpose_`. Do these belong in `CliOpts`?

**Comment by simonmichael** [comment]
Answering re the current state of things:
> Does today_ belong in ReportOpts? If so, what is it and should it be used for?
> a. Valuation date?
> b. Resolving SmartDates parsed from a query string?
> c. Resolving SmartDates parsed from -b/-e/-p
It has been there for a long time and I think is used for all of these
> There are dates involved in today_, period_, and query_. Which takes priority when all are present? Can some of these be discarded when we have the others?
https://hledger.org/hledger.html#report-start-end-date <https://hledger.org/hledger.html#report-start-end-date>  talks about how dates from -b, -e, -p, date:, date2: are combined:

"among start/end dates specified with options, the last (i.e. right-most) option takes precedence... The effective report start and end dates are the intersection of the start/end dates from options and that from date: queries."
> There's some amount of redundancy between period_ and interval_. Should we clarify the relationship?
Hmm.. is there ?  In ReportOpts, I think Interval defines the frequency/length/transitions of report subperiods, while Period defines the span of time that is the overall report period.
> A bunch of flags in ReportOpts are later put into a Query in some way, including period_, statuses_, depth_, date2_, empty_, real_, and a query string. Should we separate the query from these so it's clear which to use?
We could. Sometimes you need to check a flag, sometimes you need to apply a query, so the two forms can be handy. I think usually ReportOpts are built once and we don't need to alter them so perhaps this redundancy isn't a pressing problem ?
> There are some options which are related to display on a cli, including pretty_tables_, color_, and transpose_. Do these belong in CliOpts?
Possibly. I think some things are kept in ReportOpts because report code needs to know about them (and it can't access CliOpts). Isn't transpose one of those..

**Comment by simonmichael** [comment]
Part of the reason ReportOpts holds all these things is its job as a kind of universal "app state/reader environment" argument that we pass everywhere (in higher code, wrapped in a CliOpts/UIOpts/WebOpts). This helps keep function arguments manageable and reasonably consistent. We could instead put everything into a custom Hledger reader monad (or perhaps mess with implicit parameters - let's not go there). I usually feel that would add code complexity/difficulty that isn't really worth the benefit.

**Comment by Xitian9** [comment]
Alright, ReportSpec introduced. Making that change (especially in hledger-ui) involved changing a lot of boilerplate. If I'm going to make any other changes in there I'm sorely tempted to introduce some lenses.

**Comment by Xitian9** [comment]
Given the deep nesting of `ReportOpts` in `ReportSpec`, `CliOpts`, `UIOpts`, and `UIState`, and the fact that it's needed so often, it's probably worth it to make some lenses. We could minimise dependencies by depending only on `microlens` and `microlens-th`. We could use `makeClassy ReportOpts`, and make an instance of `HasReportOpts` for each of the above types. 

Is this something you'd be open to? We'd need to put some thougth into what the lenses would be called, but we could replace monstrosities like this:
```
-- | Toggle between showing all and showing only nonempty (more precisely, nonzero) items.
toggleEmpty :: UIState -> UIState
toggleEmpty ui@UIState{aopts=uopts@UIOpts{cliopts_=copts@CliOpts{reportspec_=rspec@ReportSpec{rsOpts=ropts}}}} =
  ui{aopts=uopts{cliopts_=copts{reportspec_=rspec{rsOpts=toggleEmpty ropts}}}}
  where
    toggleEmpty ropts = ropts{empty_=not $ empty_ ropts}
```
with the more understandable (naming of `empty_` TBD)
```
-- | Toggle between showing all and showing only nonempty (more precisely, nonzero) items.
toggleEmpty :: UIState -> UIState
toggleEmpty = over empty_ not
```

**Comment by Xitian9** [comment]
It might also be worth rethinking some of the naming here. `ReportOpts` has a trailing underscore, `ReportSpec` prefixed rs and uses camelCase, and others do some combination of these. I admit I haven't contributed to consistency with my choices here, so I'm happy to change them once we decide what to do.

**Comment by simonmichael** [comment]
Hi @Xitian9.. I just had another look at your latest, where you have implemented the new ReportSpec type. I don't have a super strong opinion on it. ReportOpts still has some dependency on the current date, but less so now. Overall this seems to clean up code a bit. Do you feel it's a good improvement overall, justifying the new type ? If so I'm ok with merging the work so far. We would just need a haddock describing the purpose of ReportSpec. Let's discuss lens later just to limit the scope.

**Comment by simonmichael** [comment]
PS I'm going to install this now and try to exercise it a bit.

**Comment by Xitian9** [comment]
Sure thing. Let me know your thoughts.

**Comment by simonmichael** [comment]
I haven't noticed any problems, I'm ready to merge this but could we have a haddock describing the purpose of ReportSpec ?

**Comment by Xitian9** [comment]
Documentation added and rebased.

**Comment by simonmichael** [comment]
Ok, thank you very much!

**Comment by simonmichael** [comment]
Found a regression in hledger-ui due to fcbe511d (ui,web: Adopt new ReportOptions interface, store literal query string in UIOpts): pressing number keys changes the depth: indicator but the displayed accounts don't change.

**Comment by simonmichael** [comment]
PS: `make ghci-ui` & :reload will be handy, if you didn't already know it.

**Comment by Xitian9** [comment]
This seems to be part of a larger suite of regressions. The same should happen when toggling Unmarked/Pending/Cleared, when filtering by real or not, or anything else should affect both ReportOpts and the query of ReportSpec. Currently it seems we're only updating the `Query` when `querystring_` is updated. I'll look to fix that.

**Comment by simonmichael** [comment]
Fixed by #1383.

**Pull request merged**